### PR TITLE
Allow multiple roles and fix project_list

### PR DESCRIPTION
**Pull request by dguerri**
This change includes:
- assignment of multiple roles for a given project;
- fix project_list when roles are assigned with this module;
- takes roles, project and domain parameters from keystone
  configuration rather than using hardcoded values.

### CODE REVIEW

File: hybrid_assignment.py [hunk 1]
```diff
@@ -31,32 +45,59 @@ def _get_metadata(self, user_id=None, tenant_id=None,
...
+            with sql.transaction() as session:
+                query = session.query(sql_assign.Role)
+                query = query.filter(sql_assign.Role.name.in_(
+                    CONF.ldap_hybrid.default_roles))
+                role_refs = query.all()
+
+            _default_roles = [role_ref.id for role_ref in role_refs]
+            if len(_default_roles) != len(CONF.ldap_hybrid.default_roles):
+                raise exception.RoleNotFound(
+                    message=_('Could not find one or more roles: %s') %
```
Comment (iartarisi) [inline hybrid_assignment.py:88]: This is missing a gettext import for _

File: hybrid_assignment.py [hunk 2]
```diff
@@ -31,32 +45,59 @@ def _get_metadata(self, user_id=None, tenant_id=None,
...
-            try:
-                role = session.query(sql_assign.Role).filter_by(
-                    name=DEFAULT_ROLE).one()
-            except sql.NotFound:
-                raise exception.RoleNotFound(role_id=DEFAULT_ROLE)
-            self._default_role = role.id
-
-        return self._default_role
+    def default_roles(self):
+        if len(self._default_roles) == 0:
```
Comment (iartarisi) [inline hybrid_assignment.py]: I think it would be better style to write `if not self._default_roles` here instead.

File: hybrid_assignment.py [hunk 3]
```diff
@@ -31,32 +45,59 @@ def _get_metadata(self, user_id=None, tenant_id=None,
...
+            with sql.transaction() as session:
+                query = session.query(sql_assign.Role)
+                query = query.filter(sql_assign.Role.name.in_(
+                    CONF.ldap_hybrid.default_roles))
+                role_refs = query.all()
+
+            _default_roles = [role_ref.id for role_ref in role_refs]
```
Comment (iartarisi) [inline hybrid_assignment.py]: maybe this line should go after the if below as you don't need the ids to raise that error.

**Comment by iartarisi** [comment]
Looks cool. I just got some time today to test this and it works fine.

There's just the import bug I'd like to see fixed before merging this. The rest are just suggestions.

**Comment by dguerri** [comment]
Thanks Ionuț. I will submit a new PR shortly.

**Comment by iartarisi** [comment]
@dguerri Cool, but please update this one rather than opening a new one ;)

**Comment by dguerri** [comment]
That should be it.
I can't try the code at the moment, so please give it a go before merging! :)

File: hybrid_assignment.py [hunk 4]
```diff
@@ -31,32 +47,59 @@ def _get_metadata(self, user_id=None, tenant_id=None,
...
+
+            self._default_roles = [role_ref.id for role_ref in role_refs]
+        return self._default_roles
+
+    def list_projects_for_user(self, user_id, group_ids, hints):
+        projects = super(Assignment, self).list_projects_for_user(
+            user_id, group_ids, hints)
+
+        # Make sure the default project is in the project list for the user
+        # user_id
```
Comment (iartarisi) [inline hybrid_assignment.py:99]: I guess this `user_id` is a superfluous word? Or am I misreading this comment?

**Comment by iartarisi** [comment]
@dguerri Thanks. Sure, I can give it a go here, but I might have to leave soon.

Could you squash the `typo` etc. commits into the bigger one and then force push to this branch?

File: hybrid_assignment.py (hunk 4 above)
Comment (dguerri) [inline hybrid_assignment.py:99]: I mean: user_id for the user

Comment (iartarisi) [inline hybrid_assignment.py:99]: Ah, ok. I get it now. Was a bit hard to parse.

Comment (dguerri) [inline hybrid_assignment.py:99]: hehe Indeed ;)

**Comment by iartarisi** [comment]
Seems to work fine. Thanks a lot!

**Comment by dguerri** [comment]
Not at all. Thank you for this module, it is extremely useful.
I think its functionalities should be integrated in Keystone. How about a new BP targeted for the K release?
I can help. Just let me know.

Cheers.

**Comment by iartarisi** [comment]
Sure, I'm open to having this merged upstream. I opened a blueprint a long time ago: https://blueprints.launchpad.net/keystone/+spec/ldap-sql-hybrid-backend but it didn't get that far. I was going to start another blueprint, but it hasn't been a priority so far. We could work on something.
